ide/che-core-ide-app/src/main/java/org/eclipse/che/ide/processes/MachineNodeRenderStrategy.java

```diff
@@ -0,0 +1,158 @@
+                                     AppContext appContext,
+                                     MachineMonitors machineMonitors) {
+        this.resources = resources;
+        this.locale = locale;
+        this.appContext = appContext;
+        this.machineMonitors = machineMonitors;
+    }
+
+    @Override
+    public SpanElement renderSpanElementFor(ProcessTreeNode candidate) {
+        return createMachineElement(candidate);
+    }
+
+    private SpanElement createMachineElement(final ProcessTreeNode node) {
+        final MachineEntity machine = (MachineEntity)node.getData();
+        final String machineId = machine.getId();
+        final MachineConfig machineConfig = machine.getConfig();
+
+        SpanElement root = Elements.createSpanElement();
+
+        Element statusElement = Elements.createSpanElement(resources.getCss().machineStatus());
+        root.appendChild(statusElement);
+
+        if (node.isRunning()) {
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusRunning()));
+        } else {
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusPausedLeft()));
+            statusElement.appendChild(Elements.createDivElement(resources.getCss().machineStatusPausedRight()));
+        }
+
+        Tooltip.create(statusElement, BOTTOM, MIDDLE, locale.viewMachineRunningTooltip());
+
+        Workspace workspace = appContext.getWorkspace();
+        if (workspace != null && RUNNING == workspace.getStatus() && node.hasTerminalAgent()) {
+            SpanElement newTerminalButton = Elements.createSpanElement(resources.getCss().newTerminalButton());
+            newTerminalButton.appendChild((Node)new SVGImage(resources.addTerminalIcon()).getElement());
+            root.appendChild(newTerminalButton);
+
+            Tooltip.create(newTerminalButton, BOTTOM, MIDDLE, locale.viewNewTerminalTooltip());
+
+            newTerminalButton.addEventListener(Event.CLICK, event -> {
+                event.stopPropagation();
+                event.preventDefault();
+
+                if (addTerminalClickHandler != null) {
+                    addTerminalClickHandler.onAddTerminalClick(machineId);
+                }
+            }, true);
+
+            EventListener blockMouseListener = event -> {
+                event.stopPropagation();
+                event.preventDefault();
+            };
+
+            newTerminalButton.addEventListener(Event.MOUSEDOWN, blockMouseListener, true);
+            newTerminalButton.addEventListener(Event.MOUSEUP, blockMouseListener, true);
+            newTerminalButton.addEventListener(Event.CLICK, blockMouseListener, true);
+            newTerminalButton.addEventListener(Event.DBLCLICK, blockMouseListener, true);
+        }
+
+        if (node.isRunning() && node.hasSSHAgent()) {
+            SpanElement sshButton = Elements.createSpanElement(resources.getCss().sshButton());
```

Comment:
I think, we should create separate method for creating new ssh element "createNewSshElement" and use it here.